### ISSUE TITLE
Bound std of Gaussian policy via beta-sigmoid ?

### ISSUE THREAD

**Issue by zuoxingdong** [rail-berkeley/softlearning#77]
If I got it correctly, the logstd of the Gaussian policy is clipped via min/max range and std is retrieved by exponentiation.

I am curious if using beta-sigmiodal function to model the logstd would be a tiny bit more stable, because it allows smooth lower/upper bound and less sharp gradient for larger magnitude.

e.g. 
```python
logvar = network output
var = 1/(1 + self.beta*torch.exp(-logvar))
var = min_var + (max_var - min_var)*var
```

**Comment by haarnoja** [comment]
The reason for clipping the std is mainly to prevent nans during initial transients, and the bounds are not typically active afterwards. You could try out different ways to implement clipping, but my intuition is that there isn't going to be much difference in performance and stability.

**Comment by zuoxingdong** [comment]
Thanks @haarnoja for your reply. I have a tiny further question, it seems the 'pre-tanh' Gaussian sampled actions is implemented by [Affine transformation of a unit Gaussian](https://github.com/rail-berkeley/softlearning/blob/master/softlearning/policies/gaussian_policy.py#L121), is it because of better numerical stability or it can be identical implementation to directly using mean and diagonal std for `MultivariateNormalDiag` ?

**Comment by hartikainen** [comment]
> it seems the 'pre-tanh' Gaussian sampled actions is implemented by [Affine transformation of a unit Gaussian](https://github.com/rail-berkeley/softlearning/blob/master/softlearning/policies/gaussian_policy.py#L121), is it because of better numerical stability or it can be identical implementation to directly using mean and diagonal std for `MultivariateNormalDiag` ?

Good question! You're right, this would be identical with directly using mean and std for the `MultivariateNormalDiag`. The reason I implemented this way was that at the time of writing these I wasn't sure what would be the best way to fit different distributions into the keras models. Also, I wanted to fit the `RealNVPPolicy` (which unfortunately doesn't yet exist in this code base) and the `GaussianPolicy` into a similar pattern to make it more readable and usable.

I don't actually think that the policies we have right now are implemented in the best possible way. Better way would be to utilize the new tensorflow probability layers to do these transformations. I'm planning to refactor them once we upgrade to tf2.

**Comment by zuoxingdong** [comment]
@hartikainen Thank you so much for your detailed explanation. I have a quick further question, in PG methods, we often stop gradients of a normally sampled action (say with reparametrization trick) when computing log-probability, otherwise there'd be zero gradient for the mean head. However, in SAC (tanh-transformed Gaussian), it seems to me that we need the gradient flow during log-abs-det part, but should we also allow gradient flow to the base distribution (i.e. original Gaussian) ?

**Comment by zuoxingdong** [comment]
Thanks @hartikainen @haarnoja for your kind replies. It is much clearer to me now. 

By the way, the numerically stable formula for log-abs-det proposed in this repo is really beautiful, and compared with clamping of original formula, it also leads to a bit better final performance. 

![image](https://user-images.githubusercontent.com/18168681/57128731-9e4dd880-6d94-11e9-88b0-4b6a76ed3873.png)

**Comment by haarnoja** [comment]
Thanks! This was actually originally proposed by @gjtucker and we have indeed found it to be really great.
